### DISCUSSION TITLE
Lock conflict between insert process to pgroonga indexed table and standby-maitenar process.

### DISCUSSION THREAD

**Discussion by isono25** [pgroonga/pgroonga#538] [Q&A]
**What happened?**
Synchronization by streaming replication in a redundant configuration of two units.
Locks remained in the insert process to the table indexed by pgroonga on the primary side and in the standby-maintenar process.
We tried to delete the standby-maitenar process with pg_cancel_backend and pg_terminate_backend, but failed, so we restarted the DB to resolve the problem.
We believe it was a lock conflict, but would like to know the cause.

```
SELECT pid,xact_start,query,state,wait_event_type FROM pg_stat_activity where state <> 'idle'
"pid","xact_start","query","state","wait_event_type"
"875756","2024-mm-dd 09:20:49.246787+09","pgroonga: standby-maintainer: maintaining","active","Lock"
"18596","2024-mm-dd 09:20:07.996925+09","insert into table_name (","active","Lock"
```

**How to reproduce it**
I ran the same process several times but could not reproduce it.

**Environment**
OS: Red Hat Enterprise Linux 8
Postgresql Version: 13.10
PGroonga Version: 3.0.7
libgroonga Version: 11.1.0

**Comment by kou** [comment]
Are you running `pgroonga_standby_maintainer` on primary?

**Reply by isono25** [reply]
Yes, I am running groonga_standby_maintainer in primary. Is this a process that is runninged so that it can continue to operate if it switches to standby after a failover, but is not normally runninged in primary?

**Reply by kou** [reply]
We should not run `pgroonga_standby_maintainer` in primary.
We should only run it in standby.

What high-availability are you using? Does it provide a hook so that we can change PostgreSQL/PGroonga configurations on failover?

**Reply by isono25** [reply]
Thank you for your response.
High availability is using pgpool-II's streaming replication mode.
For hooks, there is a function that can be shell-executed in pgpool-II.
I am thinking of considering a way to start pgroonga_stanby_maintenar with failover_command of pgpool-II at failover time,
Is there any other way to start pgroonga_stanby_maintenar at failover?

**Reply by kou** [reply]
OK. You can use the `failover_command`.
You can just remove `pgroonga_standby_maintainer` from `shared_preload_libraries` and restart when your standby becomes primary.

**Reply by isono25** [reply]
Thanks for the quick response. It was very helpful.
